File: proof_capture_backend/app/schemas/payment_proof.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.payment_proof import CaptureRequestStatus, ProofValidationStatus


class AssignCaptureUserRequest(BaseModel):
  capture_user_id: UUID | None
  capture_device_label: str | None = Field(default=None, max_length=120)


class RejectProofRequest(BaseModel):
  reason: str = Field(min_length=3, max_length=500)


class PaymentSummary(BaseModel):
  payment_id: UUID
  order_id: UUID
  branch_id: UUID
  amount: Decimal
  currency: str | None = None
  payment_method_name: str
  status: str


class CaptureRequestSummary(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  cash_session_id: UUID
  payment_id: UUID
  branch_id: UUID
  requested_by_user_id: UUID
  assigned_capture_user_id: UUID
  status: CaptureRequestStatus
  secure_token: str
  token_expires_at: datetime
  opened_at: datetime | None = None
  uploaded_at: datetime | None = None
  approved_at: datetime | None = None
  rejected_at: datetime | None = None
  canceled_at: datetime | None = None
  created_at: datetime
  updated_at: datetime


class MobileCaptureRequestView(BaseModel):
  capture_request: CaptureRequestSummary
  payment: PaymentSummary
  capture_user_name: str | None = None
  cash_session_status: str


class PaymentProofSummary(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  payment_id: UUID
  capture_request_id: UUID
  bucket_name: str
  object_path: str
  file_name_stored: str
  original_file_name: str | None = None
  mime_type: str
  file_size: int
  sha256_hash: str
  image_width: int | None = None
  image_height: int | None = None
  uploaded_by_user_id: UUID
  uploaded_at: datetime
  validation_status: ProofValidationStatus
  validated_by_user_id: UUID | None = None
  validated_at: datetime | None = None
  rejection_reason: str | None = None
  created_at: datetime
  updated_at: datetime


class PaymentProofWithRequest(BaseModel):
  proof: PaymentProofSummary | None
  capture_request: CaptureRequestSummary | None


class SignedProofUrlResponse(BaseModel):
  url: str
  expires_in_seconds: int


class PendingCaptureRequestItem(BaseModel):
  capture_request_id: UUID
  secure_token: str
  payment_id: UUID
  amount: Decimal
  payment_method_name: str
  status: CaptureRequestStatus
  requested_at: datetime

File: proof_capture_frontend_vue/src/types/payment-proof.ts
export type CaptureRequestStatus =
  | "pending"
  | "opened"
  | "uploaded"
  | "approved"
  | "rejected"
  | "expired"
  | "canceled";

export type ProofValidationStatus = "pending" | "approved" | "rejected";

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  error_code?: string | null;
}

export interface CaptureRequestSummary {
  id: string;
  cash_session_id: string;
  payment_id: string;
  branch_id: string;
  requested_by_user_id: string;
  assigned_capture_user_id: string;
  status: CaptureRequestStatus;
  secure_token: string;
  token_expires_at: string;
  opened_at?: string | null;
  uploaded_at?: string | null;
  approved_at?: string | null;
  rejected_at?: string | null;
  canceled_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PaymentSummary {
  payment_id: string;
  order_id: string;
  branch_id: string;
  amount: string;
  currency?: string | null;
  payment_method_name: string;
  status: string;
}

export interface PaymentProofSummary {
  id: string;
  payment_id: string;
  capture_request_id: string;
  bucket_name: string;
  object_path: string;
  file_name_stored: string;
  original_file_name?: string | null;
  mime_type: string;
  file_size: number;
  sha256_hash: string;
  image_width?: number | null;
  image_height?: number | null;
  uploaded_by_user_id: string;
  uploaded_at: string;
  validation_status: ProofValidationStatus;
  validated_by_user_id?: string | null;
  validated_at?: string | null;
  rejection_reason?: string | null;
  created_at: string;
  updated_at: string;
}

export interface MobileCaptureRequestView {
  capture_request: CaptureRequestSummary;
  payment: PaymentSummary;
  capture_user_name?: string | null;
  cash_session_status: string;
}

export interface PaymentProofWithRequest {
  proof?: PaymentProofSummary | null;
  capture_request?: CaptureRequestSummary | null;
}

export interface SignedProofUrlResponse {
  url: string;
  expires_in_seconds: number;
}

export interface PendingCaptureRequestItem {
  capture_request_id: string;
  secure_token: string;
  payment_id: string;
  amount: string;
  payment_method_name: string;
  status: CaptureRequestStatus;
  requested_at: string;
}

export interface CaptureUserOption {
  id: string;
  fullName: string;
  username: string;
  lastSeenAt?: string | null;
  isOnline?: boolean;
}
